lib/controllers/services.ts

```diff
@@ -0,0 +1,127 @@
+
+import { ServicePublic as ApiService } from "../api/definitions/ServicePublic";
+
+import {
+  AzureApiAuthMiddleware,
+  IAzureApiAuthorization,
+  UserGroup
+} from "../utils/middlewares/azure_api_auth";
+
+import { RequiredParamMiddleware } from "../utils/middlewares/required_param";
+
+import {
+  withRequestMiddlewares,
+  wrapRequestHandler
+} from "../utils/request_middleware";
+
+import {
+  IResponseErrorNotFound,
+  IResponseErrorQuery,
+  IResponseSuccessJson,
+  ResponseErrorNotFound,
+  ResponseErrorQuery,
+  ResponseSuccessJson
+} from "../utils/response";
+
+import { NonEmptyString } from "../utils/strings";
+
+import {
+  AzureUserAttributesMiddleware,
+  IAzureUserAttributes
+} from "../utils/middlewares/azure_user_attributes";
+
+import {
+  checkSourceIpForHandler,
+  clientIPAndCidrTuple as ipTuple
+} from "../utils/source_ip_check";
+
+type IGetServiceHandlerRet =
+  | IResponseSuccessJson<ApiService>
+  | IResponseErrorNotFound
+  | IResponseErrorQuery;
+
+type IGetServiceHandler = (
+  auth: IAzureApiAuthorization,
+  clientIp: ClientIp,
+  userAttributes: IAzureUserAttributes,
+  serviceId: NonEmptyString
+) => Promise<IGetServiceHandlerRet>;
+
+/**
+ * Converts a retrieved service to a service that can be shared via API
+ */
+function retrievedServiceToPublic(
+  retrievedService: RetrievedService
+): ApiService {
+  return {
+    department_name: retrievedService.departmentName,
+    id: retrievedService.id,
+    organization_name: retrievedService.organizationName,
+    service_id: retrievedService.serviceId,
+    service_name: retrievedService.serviceName,
+    version: retrievedService.version
+  };
+}
+
+/**
+ * Extracts the serviceId value from the URL path parameter.
+ */
+const requiredServiceIdMiddleware = RequiredParamMiddleware(
+  "serviceid",
+  NonEmptyString
+);
+
+export function GetServiceHandler(
+  serviceModel: ServiceModel
+): IGetServiceHandler {
+  return async (_, __, ___, serviceId) =>
+    (await serviceModel.findOneByServiceId(serviceId)).fold<
+      IGetServiceHandlerRet
+    >(
+      error => ResponseErrorQuery("Error while retrieving the service", error),
+      maybeService =>
+        maybeService.fold<
+          IResponseErrorNotFound | IResponseSuccessJson<ApiService>
+        >(
+          () =>
+            ResponseErrorNotFound(
+              "Service not found",
+              "The service you requested was not found in the system."
+            ),
+          service => ResponseSuccessJson(retrievedServiceToPublic(service))
+        )
+    );
+}
+
+/**
+ * Wraps a GetService handler inside an Express request handler.
+ */
+export function GetService(serviceModel: ServiceModel): express.RequestHandler {
+  const handler = GetServiceHandler(serviceModel);
+  const azureUserAttributesMiddleware = AzureUserAttributesMiddleware(
+    serviceModel
+  );
+  const middlewaresWrap = withRequestMiddlewares(
+    AzureApiAuthMiddleware(new Set([UserGroup.ApiServiceRead])),
```

Comment:
I think we should differentiate the permissions for the admin endpoint and the public endpoint, we should have an `ApiPublicServiceRead` permission